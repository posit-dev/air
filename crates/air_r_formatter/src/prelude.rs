pub(crate) use crate::{AsFormat as _, FormatNodeRule, RFormatContext, RFormatter};
pub use biome_formatter::prelude::*;
pub use biome_rowan::{AstNode as _, AstNodeList as _, AstSeparatedList as _};
