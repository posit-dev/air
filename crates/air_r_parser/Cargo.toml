[package]
authors.workspace = true
categories.workspace = true
description = "Air R parser"
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
name = "air_r_parser"
publish = false
repository.workspace = true
rust-version.workspace = true
version = "0.0.0"

[dependencies]
air_r_factory = { workspace = true }
air_r_syntax = { workspace = true }
biome_diagnostics = { workspace = true }
biome_parser = { workspace = true }
biome_rowan = { workspace = true }
biome_unicode_table = { workspace = true }
serde = { workspace = true, features = ["derive"] }
tracing = { workspace = true }
tree-sitter = { workspace = true }
tree-sitter-r = { workspace = true }

[dev-dependencies]
biome_console = { workspace = true }
biome_diagnostics = { workspace = true }
insta = { workspace = true }
line_ending = { workspace = true }
tests_macros = { workspace = true }

# cargo-workspaces metadata
[package.metadata.workspaces]
independent = true

[lints]
workspace = true
