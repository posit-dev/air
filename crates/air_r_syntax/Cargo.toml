[package]
name = "air_r_syntax"
version = "0.0.0"
authors.workspace = true
edition.workspace = true
rust-version.workspace = true
description = "SyntaxKind and common rowan definitions for air_r_parser"
homepage.workspace = true
repository.workspace = true
license.workspace = true
keywords.workspace = true
categories.workspace = true
publish = false

[dependencies]
biome_rowan = { workspace = true, features = ["serde"] }
serde = { workspace = true, features = ["derive"] }

[lints]
workspace = true
