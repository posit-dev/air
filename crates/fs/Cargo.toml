
[package]
authors.workspace = true
categories.workspace = true
description = "Air filesystem utilities"
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
name = "fs"
publish = false
repository.workspace = true
rust-version.workspace = true
version = "0.0.0"

[dependencies]
path-absolutize = { workspace = true, features = [
    "once_cell_cache",
    "use_unix_paths_on_wasm",
] }

[lints]
workspace = true
