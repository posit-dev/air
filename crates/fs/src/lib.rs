use path_absolutize::Absolutize;
use std::ffi::OsStr;
use std::path::Path;
use std::path::PathBuf;

pub fn has_r_extension(path: &Path) -> bool {
    path.extension()
        .and_then(OsStr::to_str)
        .map_or(false, is_r_extension)
}

pub fn is_r_extension(extension: &str) -> bool {
    matches!(extension, "r" | "R")
}

/// Convert any path to an absolute path (based on the current working
/// directory).
pub fn normalize_path<P: AsRef<Path>>(path: P) -> PathBuf {
    let path = path.as_ref();
    if let Ok(path) = path.absolutize() {
        path.to_path_buf()
    } else {
        path.to_path_buf()
    }
}

/// Convert an absolute path to be relative to the current working directory.
pub fn relativize_path<P: AsRef<Path>>(path: P) -> String {
    let path = path.as_ref();

    #[cfg(target_arch = "wasm32")]
    let cwd = Path::new(".");
    #[cfg(not(target_arch = "wasm32"))]
    let cwd = path_absolutize::path_dedot::CWD.as_path();

    if let Ok(path) = path.strip_prefix(cwd) {
        return format!("{}", path.display());
    }
    format!("{}", path.display())
}
