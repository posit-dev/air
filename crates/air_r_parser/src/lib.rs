mod error;
mod options;
mod parse;

#[allow(unused)]
mod treesitter;

use air_r_factory::RSyntaxFactory;
pub use error::ParseError;
pub use options::RParserOptions;
pub use parse::parse;
pub use parse::parse_r_with_cache;
pub use parse::Parse;

use air_r_syntax::RLanguage;
use biome_parser::tree_sink::LosslessTreeSink;

pub(crate) type RLosslessTreeSink<'source> = LosslessTreeSink<'source, RLanguage, RSyntaxFactory>;
