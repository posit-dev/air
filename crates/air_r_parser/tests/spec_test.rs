use air_r_parser::ParseError;
use air_r_parser::{parse, RParserOptions};
use air_r_syntax::{RRoot, RSyntaxNode};
use biome_console::fmt::{Formatter, Termcolor};
use biome_console::markup;
use biome_diagnostics::display::PrintDiagnostic;
use biome_diagnostics::termcolor;
use biome_diagnostics::DiagnosticExt;
use biome_parser::prelude::ParseDiagnostic;
use biome_rowan::SyntaxNode;
use biome_rowan::SyntaxSlot;
use biome_rowan::{AstNode, SyntaxKind};
use std::fmt::Write;
use std::fs;
use std::path::Path;

#[derive(Copy, Clone)]
pub enum ExpectedOutcome {
    Pass,
    Fail,
    Undefined,
}

#[ignore]
#[test]
fn quick_test() {
    let code = "
    1$foo
    1$2";
    let options = RParserOptions::default();
    let parsed = parse(code, options);
    let root = RRoot::unwrap_cast(parsed.syntax());
    println!("{root:#?}");
}

pub fn run(test_case: &str, _snapshot_name: &str, test_directory: &str, outcome_str: &str) {
    let outcome = match outcome_str {
        "ok" => ExpectedOutcome::Pass,
        "error" => ExpectedOutcome::Fail,
        "undefined" => ExpectedOutcome::Undefined,
        _ => panic!("Invalid expected outcome {outcome_str}"),
    };

    let test_case_path = Path::new(test_case);

    let file_name = test_case_path
        .file_name()
        .expect("Expected test to have a file name")
        .to_str()
        .expect("File name to be valid UTF8");

    let content = fs::read_to_string(test_case_path)
        .expect("Expected test path to be a readable file in UTF8 encoding");

    // Normalize to Unix line endings
    let content = line_ending::normalize(content);

    let options = RParserOptions::default();
    let parsed = parse(&content, options);
    let root = RRoot::unwrap_cast(parsed.syntax());
    let formatted_ast = format!("{:#?}", root);

    let mut snapshot = String::new();
    writeln!(snapshot, "\n## Input\n\n```R\n{content}\n```\n\n").unwrap();

    writeln!(
        snapshot,
        r#"## AST

```
{formatted_ast}
```

## CST

```
{:#?}
```
"#,
        parsed.syntax()
    )
    .unwrap();

    if parsed.has_errors() {
        let diagnostics: Vec<ParseDiagnostic> = parsed
            .errors()
            .iter()
            .map(ParseError::diagnostic)
            .map(ParseDiagnostic::clone)
            .collect();

        let mut diagnostics_buffer = termcolor::Buffer::no_color();

        let termcolor = &mut Termcolor(&mut diagnostics_buffer);
        let mut formatter = Formatter::new(termcolor);

        for diagnostic in diagnostics {
            let error = diagnostic
                .clone()
                .with_file_path(file_name)
                .with_file_source_code(&content);

            formatter
                .write_markup(markup! {
                    {PrintDiagnostic::verbose(&error)}
                })
                .expect("failed to emit diagnostic");
        }

        let formatted_diagnostics =
            std::str::from_utf8(diagnostics_buffer.as_slice()).expect("non utf8 in error buffer");

        if matches!(outcome, ExpectedOutcome::Pass) {
            panic!("Expected no errors to be present in a test case that is expected to pass but the following diagnostics are present:\n{formatted_diagnostics}")
        }

        writeln!(snapshot, "## Diagnostics\n\n```").unwrap();
        snapshot.write_str(formatted_diagnostics).unwrap();

        writeln!(snapshot, "```\n").unwrap();
    }

    match outcome {
        ExpectedOutcome::Pass => {
            let missing_required = formatted_ast.contains("missing (required)");
            if missing_required
                || parsed
                    .syntax()
                    .descendants()
                    .any(|node: RSyntaxNode| node.kind().is_bogus())
            {
                panic!("Parsed tree of a 'OK' test case should not contain any missing required children or bogus nodes: \n {formatted_ast:#?} \n\n {formatted_ast}");
            }

            let syntax = parsed.syntax();
            if has_bogus_nodes_or_empty_slots(&syntax) {
                panic!("modified tree has bogus nodes or empty slots:\n{syntax:#?} \n\n {syntax}")
            }
        }
        ExpectedOutcome::Fail => {
            if !parsed.has_errors() {
                panic!("Failing test must have diagnostics");
            }
        }
        _ => {}
    }

    insta::with_settings!({
        prepend_module_to_snapshot => false,
        snapshot_path => &test_directory,
    }, {
        insta::assert_snapshot!(file_name, snapshot);
    });
}

/// This check is used in the parser test to ensure it doesn't emit
/// bogus nodes without diagnostics.
pub fn has_bogus_nodes_or_empty_slots<L: biome_rowan::Language>(node: &SyntaxNode<L>) -> bool {
    node.descendants().any(|descendant| {
        let kind = descendant.kind();
        if kind.is_bogus() {
            return true;
        }

        if kind.is_list() {
            return descendant
                .slots()
                .any(|slot| matches!(slot, SyntaxSlot::Empty { .. }));
        }

        false
    })
}
