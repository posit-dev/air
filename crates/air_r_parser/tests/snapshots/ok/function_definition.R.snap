---
source: crates/air_r_parser/tests/spec_test.rs
expression: snapshot
---
## Input

```R
function() 1
function(a) a
function(a, b) a + b
function(
  a # important!
) a
function(...) 1
function(a, ..., b) 1
function(a = 1, ..., b = 2) a
function(..1, ..2) get("..1") + get("..2")
function(..1 = 1, ..2 = 2) get("..1") + get("..2")
function(
  x = # important!
  4
) 1

\(x, y) 1

```


## AST

```
RRoot {
    bom_token: missing (optional),
    expressions: RExpressionList [
        RFunctionDefinition {
            name: FUNCTION_KW@0..8 "function" [] [],
            parameters: RParameters {
                l_paren_token: L_PAREN@8..9 "(" [] [],
                items: RParameterList [],
                r_paren_token: R_PAREN@9..10 ")" [] [],
            },
            body: RDoubleValue {
                value_token: R_DOUBLE_LITERAL@10..12 "1" [Whitespace(" ")] [],
            },
        },
        RFunctionDefinition {
            name: FUNCTION_KW@12..21 "function" [Newline("\n")] [],
            parameters: RParameters {
                l_paren_token: L_PAREN@21..22 "(" [] [],
                items: RParameterList [
                    RParameter {
                        name: RIdentifier {
                            name_token: IDENT@22..23 "a" [] [],
                        },
                        default: missing (optional),
                    },
                ],
                r_paren_token: R_PAREN@23..24 ")" [] [],
            },
            body: RIdentifier {
                name_token: IDENT@24..26 "a" [Whitespace(" ")] [],
            },
        },
        RFunctionDefinition {
            name: FUNCTION_KW@26..35 "function" [Newline("\n")] [],
            parameters: RParameters {
                l_paren_token: L_PAREN@35..36 "(" [] [],
                items: RParameterList [
                    RParameter {
                        name: RIdentifier {
                            name_token: IDENT@36..37 "a" [] [],
                        },
                        default: missing (optional),
                    },
                    COMMA@37..38 "," [] [],
                    RParameter {
                        name: RIdentifier {
                            name_token: IDENT@38..40 "b" [Whitespace(" ")] [],
                        },
                        default: missing (optional),
                    },
                ],
                r_paren_token: R_PAREN@40..41 ")" [] [],
            },
            body: RBinaryExpression {
                left: RIdentifier {
                    name_token: IDENT@41..43 "a" [Whitespace(" ")] [],
                },
                operator: PLUS@43..45 "+" [Whitespace(" ")] [],
                right: RIdentifier {
                    name_token: IDENT@45..47 "b" [Whitespace(" ")] [],
                },
            },
        },
        RFunctionDefinition {
            name: FUNCTION_KW@47..56 "function" [Newline("\n")] [],
            parameters: RParameters {
                l_paren_token: L_PAREN@56..57 "(" [] [],
                items: RParameterList [
                    RParameter {
                        name: RIdentifier {
                            name_token: IDENT@57..74 "a" [Newline("\n"), Whitespace("  ")] [Whitespace(" "), Comments("# important!")],
                        },
                        default: missing (optional),
                    },
                ],
                r_paren_token: R_PAREN@74..76 ")" [Newline("\n")] [],
            },
            body: RIdentifier {
                name_token: IDENT@76..78 "a" [Whitespace(" ")] [],
            },
        },
        RFunctionDefinition {
            name: FUNCTION_KW@78..87 "function" [Newline("\n")] [],
            parameters: RParameters {
                l_paren_token: L_PAREN@87..88 "(" [] [],
                items: RParameterList [
                    RParameter {
                        name: RDots {
                            value_token: DOTS@88..91 "..." [] [],
                        },
                        default: missing (optional),
                    },
                ],
                r_paren_token: R_PAREN@91..92 ")" [] [],
            },
            body: RDoubleValue {
                value_token: R_DOUBLE_LITERAL@92..94 "1" [Whitespace(" ")] [],
            },
        },
        RFunctionDefinition {
            name: FUNCTION_KW@94..103 "function" [Newline("\n")] [],
            parameters: RParameters {
                l_paren_token: L_PAREN@103..104 "(" [] [],
                items: RParameterList [
                    RParameter {
                        name: RIdentifier {
                            name_token: IDENT@104..105 "a" [] [],
                        },
                        default: missing (optional),
                    },
                    COMMA@105..106 "," [] [],
                    RParameter {
                        name: RDots {
                            value_token: DOTS@106..110 "..." [Whitespace(" ")] [],
                        },
                        default: missing (optional),
                    },
                    COMMA@110..111 "," [] [],
                    RParameter {
                        name: RIdentifier {
                            name_token: IDENT@111..113 "b" [Whitespace(" ")] [],
                        },
                        default: missing (optional),
                    },
                ],
                r_paren_token: R_PAREN@113..114 ")" [] [],
            },
            body: RDoubleValue {
                value_token: R_DOUBLE_LITERAL@114..116 "1" [Whitespace(" ")] [],
            },
        },
        RFunctionDefinition {
            name: FUNCTION_KW@116..125 "function" [Newline("\n")] [],
            parameters: RParameters {
                l_paren_token: L_PAREN@125..126 "(" [] [],
                items: RParameterList [
                    RParameter {
                        name: RIdentifier {
                            name_token: IDENT@126..127 "a" [] [],
                        },
                        default: RParameterDefault {
                            eq_token: EQUAL@127..129 "=" [Whitespace(" ")] [],
                            value: RDoubleValue {
                                value_token: R_DOUBLE_LITERAL@129..131 "1" [Whitespace(" ")] [],
                            },
                        },
                    },
                    COMMA@131..132 "," [] [],
                    RParameter {
                        name: RDots {
                            value_token: DOTS@132..136 "..." [Whitespace(" ")] [],
                        },
                        default: missing (optional),
                    },
                    COMMA@136..137 "," [] [],
                    RParameter {
                        name: RIdentifier {
                            name_token: IDENT@137..139 "b" [Whitespace(" ")] [],
                        },
                        default: RParameterDefault {
                            eq_token: EQUAL@139..141 "=" [Whitespace(" ")] [],
                            value: RDoubleValue {
                                value_token: R_DOUBLE_LITERAL@141..143 "2" [Whitespace(" ")] [],
                            },
                        },
                    },
                ],
                r_paren_token: R_PAREN@143..144 ")" [] [],
            },
            body: RIdentifier {
                name_token: IDENT@144..146 "a" [Whitespace(" ")] [],
            },
        },
        RFunctionDefinition {
            name: FUNCTION_KW@146..155 "function" [Newline("\n")] [],
            parameters: RParameters {
                l_paren_token: L_PAREN@155..156 "(" [] [],
                items: RParameterList [
                    RParameter {
                        name: RDotDotI {
                            value_token: DOTDOTI@156..159 "..1" [] [],
                        },
                        default: missing (optional),
                    },
                    COMMA@159..160 "," [] [],
                    RParameter {
                        name: RDotDotI {
                            value_token: DOTDOTI@160..164 "..2" [Whitespace(" ")] [],
                        },
                        default: missing (optional),
                    },
                ],
                r_paren_token: R_PAREN@164..165 ")" [] [],
            },
            body: RBinaryExpression {
                left: RCall {
                    function: RIdentifier {
                        name_token: IDENT@165..169 "get" [Whitespace(" ")] [],
                    },
                    arguments: RCallArguments {
                        l_paren_token: L_PAREN@169..170 "(" [] [],
                        items: RArgumentList [
                            RArgument {
                                name_clause: missing (optional),
                                value: RStringValue {
                                    value_token: R_STRING_LITERAL@170..175 "\"..1\"" [] [],
                                },
                            },
                        ],
                        r_paren_token: R_PAREN@175..176 ")" [] [],
                    },
                },
                operator: PLUS@176..178 "+" [Whitespace(" ")] [],
                right: RCall {
                    function: RIdentifier {
                        name_token: IDENT@178..182 "get" [Whitespace(" ")] [],
                    },
                    arguments: RCallArguments {
                        l_paren_token: L_PAREN@182..183 "(" [] [],
                        items: RArgumentList [
                            RArgument {
                                name_clause: missing (optional),
                                value: RStringValue {
                                    value_token: R_STRING_LITERAL@183..188 "\"..2\"" [] [],
                                },
                            },
                        ],
                        r_paren_token: R_PAREN@188..189 ")" [] [],
                    },
                },
            },
        },
        RFunctionDefinition {
            name: FUNCTION_KW@189..198 "function" [Newline("\n")] [],
            parameters: RParameters {
                l_paren_token: L_PAREN@198..199 "(" [] [],
                items: RParameterList [
                    RParameter {
                        name: RDotDotI {
                            value_token: DOTDOTI@199..202 "..1" [] [],
                        },
                        default: RParameterDefault {
                            eq_token: EQUAL@202..204 "=" [Whitespace(" ")] [],
                            value: RDoubleValue {
                                value_token: R_DOUBLE_LITERAL@204..206 "1" [Whitespace(" ")] [],
                            },
                        },
                    },
                    COMMA@206..207 "," [] [],
                    RParameter {
                        name: RDotDotI {
                            value_token: DOTDOTI@207..211 "..2" [Whitespace(" ")] [],
                        },
                        default: RParameterDefault {
                            eq_token: EQUAL@211..213 "=" [Whitespace(" ")] [],
                            value: RDoubleValue {
                                value_token: R_DOUBLE_LITERAL@213..215 "2" [Whitespace(" ")] [],
                            },
                        },
                    },
                ],
                r_paren_token: R_PAREN@215..216 ")" [] [],
            },
            body: RBinaryExpression {
                left: RCall {
                    function: RIdentifier {
                        name_token: IDENT@216..220 "get" [Whitespace(" ")] [],
                    },
                    arguments: RCallArguments {
                        l_paren_token: L_PAREN@220..221 "(" [] [],
                        items: RArgumentList [
                            RArgument {
                                name_clause: missing (optional),
                                value: RStringValue {
                                    value_token: R_STRING_LITERAL@221..226 "\"..1\"" [] [],
                                },
                            },
                        ],
                        r_paren_token: R_PAREN@226..227 ")" [] [],
                    },
                },
                operator: PLUS@227..229 "+" [Whitespace(" ")] [],
                right: RCall {
                    function: RIdentifier {
                        name_token: IDENT@229..233 "get" [Whitespace(" ")] [],
                    },
                    arguments: RCallArguments {
                        l_paren_token: L_PAREN@233..234 "(" [] [],
                        items: RArgumentList [
                            RArgument {
                                name_clause: missing (optional),
                                value: RStringValue {
                                    value_token: R_STRING_LITERAL@234..239 "\"..2\"" [] [],
                                },
                            },
                        ],
                        r_paren_token: R_PAREN@239..240 ")" [] [],
                    },
                },
            },
        },
        RFunctionDefinition {
            name: FUNCTION_KW@240..249 "function" [Newline("\n")] [],
            parameters: RParameters {
                l_paren_token: L_PAREN@249..250 "(" [] [],
                items: RParameterList [
                    RParameter {
                        name: RIdentifier {
                            name_token: IDENT@250..254 "x" [Newline("\n"), Whitespace("  ")] [],
                        },
                        default: RParameterDefault {
                            eq_token: EQUAL@254..269 "=" [Whitespace(" ")] [Whitespace(" "), Comments("# important!")],
                            value: RDoubleValue {
                                value_token: R_DOUBLE_LITERAL@269..273 "4" [Newline("\n"), Whitespace("  ")] [],
                            },
                        },
                    },
                ],
                r_paren_token: R_PAREN@273..275 ")" [Newline("\n")] [],
            },
            body: RDoubleValue {
                value_token: R_DOUBLE_LITERAL@275..277 "1" [Whitespace(" ")] [],
            },
        },
        RFunctionDefinition {
            name: BACKSLASH@277..280 "\\" [Newline("\n"), Newline("\n")] [],
            parameters: RParameters {
                l_paren_token: L_PAREN@280..281 "(" [] [],
                items: RParameterList [
                    RParameter {
                        name: RIdentifier {
                            name_token: IDENT@281..282 "x" [] [],
                        },
                        default: missing (optional),
                    },
                    COMMA@282..283 "," [] [],
                    RParameter {
                        name: RIdentifier {
                            name_token: IDENT@283..285 "y" [Whitespace(" ")] [],
                        },
                        default: missing (optional),
                    },
                ],
                r_paren_token: R_PAREN@285..286 ")" [] [],
            },
            body: RDoubleValue {
                value_token: R_DOUBLE_LITERAL@286..288 "1" [Whitespace(" ")] [],
            },
        },
    ],
    eof_token: EOF@288..289 "" [Newline("\n")] [],
}
```

## CST

```
0: R_ROOT@0..289
  0: (empty)
  1: R_EXPRESSION_LIST@0..288
    0: R_FUNCTION_DEFINITION@0..12
      0: FUNCTION_KW@0..8 "function" [] []
      1: R_PARAMETERS@8..10
        0: L_PAREN@8..9 "(" [] []
        1: R_PARAMETER_LIST@9..9
        2: R_PAREN@9..10 ")" [] []
      2: R_DOUBLE_VALUE@10..12
        0: R_DOUBLE_LITERAL@10..12 "1" [Whitespace(" ")] []
    1: R_FUNCTION_DEFINITION@12..26
      0: FUNCTION_KW@12..21 "function" [Newline("\n")] []
      1: R_PARAMETERS@21..24
        0: L_PAREN@21..22 "(" [] []
        1: R_PARAMETER_LIST@22..23
          0: R_PARAMETER@22..23
            0: R_IDENTIFIER@22..23
              0: IDENT@22..23 "a" [] []
            1: (empty)
        2: R_PAREN@23..24 ")" [] []
      2: R_IDENTIFIER@24..26
        0: IDENT@24..26 "a" [Whitespace(" ")] []
    2: R_FUNCTION_DEFINITION@26..47
      0: FUNCTION_KW@26..35 "function" [Newline("\n")] []
      1: R_PARAMETERS@35..41
        0: L_PAREN@35..36 "(" [] []
        1: R_PARAMETER_LIST@36..40
          0: R_PARAMETER@36..37
            0: R_IDENTIFIER@36..37
              0: IDENT@36..37 "a" [] []
            1: (empty)
          1: COMMA@37..38 "," [] []
          2: R_PARAMETER@38..40
            0: R_IDENTIFIER@38..40
              0: IDENT@38..40 "b" [Whitespace(" ")] []
            1: (empty)
        2: R_PAREN@40..41 ")" [] []
      2: R_BINARY_EXPRESSION@41..47
        0: R_IDENTIFIER@41..43
          0: IDENT@41..43 "a" [Whitespace(" ")] []
        1: PLUS@43..45 "+" [Whitespace(" ")] []
        2: R_IDENTIFIER@45..47
          0: IDENT@45..47 "b" [Whitespace(" ")] []
    3: R_FUNCTION_DEFINITION@47..78
      0: FUNCTION_KW@47..56 "function" [Newline("\n")] []
      1: R_PARAMETERS@56..76
        0: L_PAREN@56..57 "(" [] []
        1: R_PARAMETER_LIST@57..74
          0: R_PARAMETER@57..74
            0: R_IDENTIFIER@57..74
              0: IDENT@57..74 "a" [Newline("\n"), Whitespace("  ")] [Whitespace(" "), Comments("# important!")]
            1: (empty)
        2: R_PAREN@74..76 ")" [Newline("\n")] []
      2: R_IDENTIFIER@76..78
        0: IDENT@76..78 "a" [Whitespace(" ")] []
    4: R_FUNCTION_DEFINITION@78..94
      0: FUNCTION_KW@78..87 "function" [Newline("\n")] []
      1: R_PARAMETERS@87..92
        0: L_PAREN@87..88 "(" [] []
        1: R_PARAMETER_LIST@88..91
          0: R_PARAMETER@88..91
            0: R_DOTS@88..91
              0: DOTS@88..91 "..." [] []
            1: (empty)
        2: R_PAREN@91..92 ")" [] []
      2: R_DOUBLE_VALUE@92..94
        0: R_DOUBLE_LITERAL@92..94 "1" [Whitespace(" ")] []
    5: R_FUNCTION_DEFINITION@94..116
      0: FUNCTION_KW@94..103 "function" [Newline("\n")] []
      1: R_PARAMETERS@103..114
        0: L_PAREN@103..104 "(" [] []
        1: R_PARAMETER_LIST@104..113
          0: R_PARAMETER@104..105
            0: R_IDENTIFIER@104..105
              0: IDENT@104..105 "a" [] []
            1: (empty)
          1: COMMA@105..106 "," [] []
          2: R_PARAMETER@106..110
            0: R_DOTS@106..110
              0: DOTS@106..110 "..." [Whitespace(" ")] []
            1: (empty)
          3: COMMA@110..111 "," [] []
          4: R_PARAMETER@111..113
            0: R_IDENTIFIER@111..113
              0: IDENT@111..113 "b" [Whitespace(" ")] []
            1: (empty)
        2: R_PAREN@113..114 ")" [] []
      2: R_DOUBLE_VALUE@114..116
        0: R_DOUBLE_LITERAL@114..116 "1" [Whitespace(" ")] []
    6: R_FUNCTION_DEFINITION@116..146
      0: FUNCTION_KW@116..125 "function" [Newline("\n")] []
      1: R_PARAMETERS@125..144
        0: L_PAREN@125..126 "(" [] []
        1: R_PARAMETER_LIST@126..143
          0: R_PARAMETER@126..131
            0: R_IDENTIFIER@126..127
              0: IDENT@126..127 "a" [] []
            1: R_PARAMETER_DEFAULT@127..131
              0: EQUAL@127..129 "=" [Whitespace(" ")] []
              1: R_DOUBLE_VALUE@129..131
                0: R_DOUBLE_LITERAL@129..131 "1" [Whitespace(" ")] []
          1: COMMA@131..132 "," [] []
          2: R_PARAMETER@132..136
            0: R_DOTS@132..136
              0: DOTS@132..136 "..." [Whitespace(" ")] []
            1: (empty)
          3: COMMA@136..137 "," [] []
          4: R_PARAMETER@137..143
            0: R_IDENTIFIER@137..139
              0: IDENT@137..139 "b" [Whitespace(" ")] []
            1: R_PARAMETER_DEFAULT@139..143
              0: EQUAL@139..141 "=" [Whitespace(" ")] []
              1: R_DOUBLE_VALUE@141..143
                0: R_DOUBLE_LITERAL@141..143 "2" [Whitespace(" ")] []
        2: R_PAREN@143..144 ")" [] []
      2: R_IDENTIFIER@144..146
        0: IDENT@144..146 "a" [Whitespace(" ")] []
    7: R_FUNCTION_DEFINITION@146..189
      0: FUNCTION_KW@146..155 "function" [Newline("\n")] []
      1: R_PARAMETERS@155..165
        0: L_PAREN@155..156 "(" [] []
        1: R_PARAMETER_LIST@156..164
          0: R_PARAMETER@156..159
            0: R_DOT_DOT_I@156..159
              0: DOTDOTI@156..159 "..1" [] []
            1: (empty)
          1: COMMA@159..160 "," [] []
          2: R_PARAMETER@160..164
            0: R_DOT_DOT_I@160..164
              0: DOTDOTI@160..164 "..2" [Whitespace(" ")] []
            1: (empty)
        2: R_PAREN@164..165 ")" [] []
      2: R_BINARY_EXPRESSION@165..189
        0: R_CALL@165..176
          0: R_IDENTIFIER@165..169
            0: IDENT@165..169 "get" [Whitespace(" ")] []
          1: R_CALL_ARGUMENTS@169..176
            0: L_PAREN@169..170 "(" [] []
            1: R_ARGUMENT_LIST@170..175
              0: R_ARGUMENT@170..175
                0: (empty)
                1: R_STRING_VALUE@170..175
                  0: R_STRING_LITERAL@170..175 "\"..1\"" [] []
            2: R_PAREN@175..176 ")" [] []
        1: PLUS@176..178 "+" [Whitespace(" ")] []
        2: R_CALL@178..189
          0: R_IDENTIFIER@178..182
            0: IDENT@178..182 "get" [Whitespace(" ")] []
          1: R_CALL_ARGUMENTS@182..189
            0: L_PAREN@182..183 "(" [] []
            1: R_ARGUMENT_LIST@183..188
              0: R_ARGUMENT@183..188
                0: (empty)
                1: R_STRING_VALUE@183..188
                  0: R_STRING_LITERAL@183..188 "\"..2\"" [] []
            2: R_PAREN@188..189 ")" [] []
    8: R_FUNCTION_DEFINITION@189..240
      0: FUNCTION_KW@189..198 "function" [Newline("\n")] []
      1: R_PARAMETERS@198..216
        0: L_PAREN@198..199 "(" [] []
        1: R_PARAMETER_LIST@199..215
          0: R_PARAMETER@199..206
            0: R_DOT_DOT_I@199..202
              0: DOTDOTI@199..202 "..1" [] []
            1: R_PARAMETER_DEFAULT@202..206
              0: EQUAL@202..204 "=" [Whitespace(" ")] []
              1: R_DOUBLE_VALUE@204..206
                0: R_DOUBLE_LITERAL@204..206 "1" [Whitespace(" ")] []
          1: COMMA@206..207 "," [] []
          2: R_PARAMETER@207..215
            0: R_DOT_DOT_I@207..211
              0: DOTDOTI@207..211 "..2" [Whitespace(" ")] []
            1: R_PARAMETER_DEFAULT@211..215
              0: EQUAL@211..213 "=" [Whitespace(" ")] []
              1: R_DOUBLE_VALUE@213..215
                0: R_DOUBLE_LITERAL@213..215 "2" [Whitespace(" ")] []
        2: R_PAREN@215..216 ")" [] []
      2: R_BINARY_EXPRESSION@216..240
        0: R_CALL@216..227
          0: R_IDENTIFIER@216..220
            0: IDENT@216..220 "get" [Whitespace(" ")] []
          1: R_CALL_ARGUMENTS@220..227
            0: L_PAREN@220..221 "(" [] []
            1: R_ARGUMENT_LIST@221..226
              0: R_ARGUMENT@221..226
                0: (empty)
                1: R_STRING_VALUE@221..226
                  0: R_STRING_LITERAL@221..226 "\"..1\"" [] []
            2: R_PAREN@226..227 ")" [] []
        1: PLUS@227..229 "+" [Whitespace(" ")] []
        2: R_CALL@229..240
          0: R_IDENTIFIER@229..233
            0: IDENT@229..233 "get" [Whitespace(" ")] []
          1: R_CALL_ARGUMENTS@233..240
            0: L_PAREN@233..234 "(" [] []
            1: R_ARGUMENT_LIST@234..239
              0: R_ARGUMENT@234..239
                0: (empty)
                1: R_STRING_VALUE@234..239
                  0: R_STRING_LITERAL@234..239 "\"..2\"" [] []
            2: R_PAREN@239..240 ")" [] []
    9: R_FUNCTION_DEFINITION@240..277
      0: FUNCTION_KW@240..249 "function" [Newline("\n")] []
      1: R_PARAMETERS@249..275
        0: L_PAREN@249..250 "(" [] []
        1: R_PARAMETER_LIST@250..273
          0: R_PARAMETER@250..273
            0: R_IDENTIFIER@250..254
              0: IDENT@250..254 "x" [Newline("\n"), Whitespace("  ")] []
            1: R_PARAMETER_DEFAULT@254..273
              0: EQUAL@254..269 "=" [Whitespace(" ")] [Whitespace(" "), Comments("# important!")]
              1: R_DOUBLE_VALUE@269..273
                0: R_DOUBLE_LITERAL@269..273 "4" [Newline("\n"), Whitespace("  ")] []
        2: R_PAREN@273..275 ")" [Newline("\n")] []
      2: R_DOUBLE_VALUE@275..277
        0: R_DOUBLE_LITERAL@275..277 "1" [Whitespace(" ")] []
    10: R_FUNCTION_DEFINITION@277..288
      0: BACKSLASH@277..280 "\\" [Newline("\n"), Newline("\n")] []
      1: R_PARAMETERS@280..286
        0: L_PAREN@280..281 "(" [] []
        1: R_PARAMETER_LIST@281..285
          0: R_PARAMETER@281..282
            0: R_IDENTIFIER@281..282
              0: IDENT@281..282 "x" [] []
            1: (empty)
          1: COMMA@282..283 "," [] []
          2: R_PARAMETER@283..285
            0: R_IDENTIFIER@283..285
              0: IDENT@283..285 "y" [Whitespace(" ")] []
            1: (empty)
        2: R_PAREN@285..286 ")" [] []
      2: R_DOUBLE_VALUE@286..288
        0: R_DOUBLE_LITERAL@286..288 "1" [Whitespace(" ")] []
  2: EOF@288..289 "" [Newline("\n")] []

```
